polymetis/polymetis/src/polymetis_server.cpp

```diff
@@ -70,6 +70,52 @@ Status PolymetisControllerServerImpl::GetRobotStateLog(
   return Status::OK;
 }
 
+Status PolymetisControllerServerImpl::SetMirrorRobotState(
+    ServerContext *context, const RobotState *robot_state, Empty *) {
+  if (!robot_client_context_.metadata.is_sim()) {
+    return Status(
+        StatusCode::FAILED_PRECONDITION,
+        "SetMirrorRobotState may only be used with a simulator client!");
+  }
+  std::lock_guard<std::mutex> guard(mirror_state_mtx_);
+  auto curr_sec = mirror_sim_robot_state_.timestamp().seconds();
+  auto curr_nano = mirror_sim_robot_state_.timestamp().nanos();
+  auto rs_sec = robot_state->timestamp().seconds();
+  auto rs_nano = robot_state->timestamp().nanos();
+  if (!is_mirror_state_set_ || (curr_sec == rs_sec && curr_nano < rs_nano) ||
+      curr_sec < rs_sec) {
```

Comment:
You can simply do [`Timestamp.Equals`](https://developers.google.com/protocol-buffers/docs/reference/csharp/class/google/protobuf/well-known-types/timestamp#class_google_1_1_protobuf_1_1_well_known_types_1_1_timestamp_1a00839bf69f69d66efc0bfdd873557c7c).

Reply:
Honestly I wouldn't even check the timestamp on the server side. The user client can perform this check and thus avoid wasting a request. The server can simply update `is_mirror_state_set_` as the time & computational effort it takes to perform the assignment is trivial compared to sending and receiving the request.

Reply:
Yeah, this is true -- if we are assuming a single producer which only calls SetMirrorRobotState as blocking RPCs, then the requests are never going to be out of order.